gpsrenda/widgets/widgets.py

```diff
@@ -0,0 +1,192 @@
+from gpsrenda.widgets import *
+from gpsrenda.utils import timestamp_to_seconds, km_to_mi, c_to_f
+from gpsrenda.widgets.utils import latlondist
+
+STYLE_TABLE = {
+    'hbar': GaugeHorizontal,
+    'vbar': GaugeVertical,
+    'map': GaugeMap,
+    'grade': GaugeElevationMap,
+}
+
+class SpeedWidget:
+    def __init__(self, data_source, x, y, style='hbar', units='metric'):
+        self.data_source = data_source
+        self.units = units
+        gauge_class = STYLE_TABLE[style]
+        caption = 'mph' if units == 'imperial' else 'kph'
+
+        colors = [(0, [0.2, 0.0, 0.6]),
```

Comment:
Having the data range be configurable I found was kind of important.  For instance, I have a very different range of reasonable speeds for a century vs. crit riding.  Same for powers and heart rates, for that matter.

I'm kind of of two minds whether heart rate range (and speed range, and cadence range, and power range...) is a `widget` property or a `global` property.  On the one hand, it is specific to the widget.  On the other hand, there could be multiple heart rate gauges of different types somehow (one textual somewhere on screen, one circular or something?), and it might be nice to share ranges between all of them?

Reply:
In the back of my mind I see the need for another patch to streamline the `Widget` system.

- Cleaner unit conversions
- Propagation of more settings from widgets to gauges including colors, sizes, and text style (see your dropshadow comments) with good defaults
- Cleaner access to raw fit file data for gauges that pre-render

For now I left the widget-gauge interface in this tenuous situation because a) it works and b) it isolates the sketchy stuff to the widgets. Fit file access and movie creation are very stable.

Multiple widgets with the same info is a very uncommon use case, and it's not so hard to type something twice in the config for these case.